File: src/services/validation-engine.ts
import type {
  ExtensionSettings,
  ResxFamily,
  ResxFile,
  ResourceRow,
  ValidationIssue,
  ValidationRulesConfig,
} from '../models/types';
import { NEUTRAL_LOCALE } from '../models/types';
import {
  endingsMatch,
  placeholdersMatch,
  toPascalCaseKey,
} from './naming';

export function validateFamily(
  family: ResxFamily,
  files: ResxFile[],
  rules: ValidationRulesConfig,
  localesToCheck?: string[]
): ValidationIssue[] {
  const issues: ValidationIssue[] = [];
  const byLocale = new Map<string, ResxFile>();
  for (const f of files) {
    byLocale.set(f.locale, f);
  }

  const neutral = byLocale.get(NEUTRAL_LOCALE) ?? files[0];
  if (!neutral) {
    return issues;
  }

  if (rules.duplicateKeys) {
    for (const f of files) {
      for (const key of f.duplicateKeys) {
        issues.push({
          rule: 'duplicateKeys',
          severity: 'error',
          message: `Duplicate key "${key}" in ${localeLabel(f.locale)}`,
          key,
          locale: f.locale,
          familyId: family.id,
        });
      }
    }
  }

  const keys = new Set<string>();
  for (const e of neutral.entries) {
    keys.add(e.key);
  }
  // Also include keys that only exist in satellites
  for (const f of files) {
    for (const e of f.entries) {
      keys.add(e.key);
    }
  }

  const locales =
    localesToCheck && localesToCheck.length > 0
      ? localesToCheck
      : [...byLocale.keys()];

  for (const key of keys) {
    const neutralEntry = neutral.entries.find((e) => e.key === key);
    const neutralValue = neutralEntry?.value ?? '';

    if (rules.keyPascalCase && neutralEntry) {
      const expected = toPascalCaseKey(neutralValue);
      if (expected && key !== expected) {
        issues.push({
          rule: 'keyPascalCase',
          severity: 'warning',
          message: `Key should be PascalCase of neutral value: expected "${expected}"`,
          key,
          familyId: family.id,
        });
      }
    }

    for (const locale of locales) {
      if (locale === NEUTRAL_LOCALE && byLocale.get(NEUTRAL_LOCALE) === neutral) {
        // still validate satellites primarily for suffix/placeholders/missing
      }
      const file = byLocale.get(locale);
      const entry = file?.entries.find((e) => e.key === key);
      const value = entry?.value ?? '';

      if (locale !== (neutral.locale ?? NEUTRAL_LOCALE) || file !== neutral) {
        if (rules.missingTranslation && neutralEntry && !value) {
          // Only flag missing if this locale file exists in the family
          if (family.files[locale] !== undefined) {
            issues.push({
              rule: 'missingTranslation',
              severity: 'warning',
              message: `Missing translation for ${localeLabel(locale)}`,
              key,
              locale,
              familyId: family.id,
            });
          }
        }

        if (value && neutralValue) {
          if (rules.matchingSuffix && !endingsMatch(neutralValue, value)) {
            issues.push({
              rule: 'matchingSuffix',
              severity: 'warning',
              message: `Ending does not match neutral for ${localeLabel(locale)}`,
              key,
              locale,
              familyId: family.id,
            });
          }
          if (rules.placeholders && !placeholdersMatch(neutralValue, value)) {
            issues.push({
              rule: 'placeholders',
              severity: 'warning',
              message: `Placeholders differ from neutral for ${localeLabel(locale)}`,
              key,
              locale,
              familyId: family.id,
            });
          }
        }
      }
    }
  }

  return issues;
}

export function attachIssuesToRows(
  rows: ResourceRow[],
  issues: ValidationIssue[]
): ResourceRow[] {
  const byKey = new Map<string, ValidationIssue[]>();
  for (const issue of issues) {
    const mapKey = `${issue.familyId}::${issue.key}`;
    let list = byKey.get(mapKey);
    if (!list) {
      list = [];
      byKey.set(mapKey, list);
    }
    list.push(issue);
  }

  return rows.map((row) => ({
    ...row,
    issues: byKey.get(`${row.familyId}::${row.key}`) ?? [],
  }));
}

export function buildRows(
  family: ResxFamily,
  files: ResxFile[]
): ResourceRow[] {
  const keys = new Map<string, { comment: string; values: Record<string, string> }>();

  for (const file of files) {
    for (const entry of file.entries) {
      let row = keys.get(entry.key);
      if (!row) {
        row = { comment: '', values: {} };
        keys.set(entry.key, row);
      }
      row.values[file.locale] = entry.value;
      if (file.locale === NEUTRAL_LOCALE && entry.comment) {
        row.comment = entry.comment;
      } else if (!row.comment && entry.comment) {
        row.comment = entry.comment;
      }
    }
  }

  const rows: ResourceRow[] = [];
  for (const [key, data] of keys) {
    rows.push({
      familyId: family.id,
      key,
      comment: data.comment,
      values: data.values,
      issues: [],
    });
  }

  rows.sort((a, b) => a.key.localeCompare(b.key, undefined, { sensitivity: 'base' }));
  return rows;
}

export function defaultSettings(): ExtensionSettings {
  return {
    neutralLocale: '',
    keyNaming: 'pascalFromNeutral',
    updateDesignerCs: true,
    visibleLocales: [],
    rules: {
      keyPascalCase: true,
      matchingSuffix: true,
      placeholders: true,
      missingTranslation: true,
      duplicateKeys: true,
    },
  };
}

function localeLabel(locale: string): string {
  return locale === NEUTRAL_LOCALE || locale === '' ? 'Neutral' : locale;
}

File: apps/visualstudio/ResXGuard.Core/Services/ValidationEngine.cs
namespace ResXGuard.Core;

public static class ValidationEngine
{
    public static List<ValidationIssue> ValidateFamily(
        ResxFamily family,
        IReadOnlyList<ResxFile> files,
        ValidationRulesConfig rules,
        IReadOnlyList<string>? localesToCheck = null)
    {
        var issues = new List<ValidationIssue>();
        var byLocale = files.ToDictionary(f => f.Locale, f => f);

        if (!byLocale.TryGetValue(ResxConstants.NeutralLocale, out var neutral))
            neutral = files.FirstOrDefault();
        if (neutral == null) return issues;

        if (rules.DuplicateKeys)
        {
            foreach (var f in files)
            {
                foreach (var key in f.DuplicateKeys)
                {
                    issues.Add(new ValidationIssue
                    {
                        Rule = IssueRule.DuplicateKeys,
                        Severity = IssueSeverity.Error,
                        Message = $"Duplicate key \"{key}\" in {LocaleLabel(f.Locale)}",
                        Key = key,
                        Locale = f.Locale,
                        FamilyId = family.Id
                    });
                }
            }
        }

        var keys = new HashSet<string>(StringComparer.Ordinal);
        foreach (var e in neutral.Entries) keys.Add(e.Key);
        foreach (var f in files)
            foreach (var e in f.Entries) keys.Add(e.Key);

        var locales = localesToCheck?.Count > 0
            ? localesToCheck.ToList()
            : family.Files.Keys.Union(byLocale.Keys).Distinct().ToList();

        foreach (var key in keys)
        {
            var neutralEntry = neutral.Entries.FirstOrDefault(e => e.Key == key);
            var neutralValue = neutralEntry?.Value ?? "";

            if (rules.KeyPascalCase && neutralEntry != null)
            {
                var expected = Naming.ToPascalCaseKey(neutralValue);
                if (!string.IsNullOrEmpty(expected) && key != expected)
                {
                    issues.Add(new ValidationIssue
                    {
                        Rule = IssueRule.KeyPascalCase,
                        Severity = IssueSeverity.Warning,
                        Message = $"Key should be PascalCase of neutral value: expected \"{expected}\"",
                        Key = key,
                        FamilyId = family.Id,
                        SuggestedKey = expected
                    });
                }
            }

            foreach (var locale in locales)
            {
                byLocale.TryGetValue(locale, out var file);
                var entry = file?.Entries.FirstOrDefault(e => e.Key == key);
                var value = entry?.Value ?? "";

                if (locale != neutral.Locale || file != neutral)
                {
                    if (rules.MissingTranslation && neutralEntry != null && string.IsNullOrEmpty(value))
                    {
                        if (family.Files.ContainsKey(locale))
                        {
                            issues.Add(new ValidationIssue
                            {
                                Rule = IssueRule.MissingTranslation,
                                Severity = IssueSeverity.Warning,
                                Message = $"Missing translation for {LocaleLabel(locale)}",
                                Key = key,
                                Locale = locale,
                                FamilyId = family.Id
                            });
                        }
                    }

                    if (!string.IsNullOrEmpty(value) && !string.IsNullOrEmpty(neutralValue))
                    {
                        if (rules.MatchingSuffix && !Naming.EndingsMatch(neutralValue, value))
                        {
                            issues.Add(new ValidationIssue
                            {
                                Rule = IssueRule.MatchingSuffix,
                                Severity = IssueSeverity.Warning,
                                Message = $"Ending does not match neutral for {LocaleLabel(locale)}",
                                Key = key,
                                Locale = locale,
                                FamilyId = family.Id
                            });
                        }
                        if (rules.Placeholders && !Naming.PlaceholdersMatch(neutralValue, value))
                        {
                            issues.Add(new ValidationIssue
                            {
                                Rule = IssueRule.Placeholders,
                                Severity = IssueSeverity.Warning,
                                Message = $"Placeholders differ from neutral for {LocaleLabel(locale)}",
                                Key = key,
                                Locale = locale,
                                FamilyId = family.Id
                            });
                        }
                    }
                }
            }
        }
        return issues;
    }

    public static List<ResourceRow> AttachIssuesToRows(IReadOnlyList<ResourceRow> rows, IReadOnlyList<ValidationIssue> issues)
    {
        var byKey = new Dictionary<string, List<ValidationIssue>>();
        foreach (var issue in issues)
        {
            var mapKey = $"{issue.FamilyId}::{issue.Key}";
            if (!byKey.TryGetValue(mapKey, out var list))
            {
                list = new List<ValidationIssue>();
                byKey[mapKey] = list;
            }
            list.Add(issue);
        }
        return rows.Select(row => new ResourceRow
        {
            FamilyId = row.FamilyId,
            Key = row.Key,
            Comment = row.Comment,
            Values = new Dictionary<string, string>(row.Values),
            Issues = byKey.TryGetValue($"{row.FamilyId}::{row.Key}", out var list) ? list : new List<ValidationIssue>()
        }).ToList();
    }

    public static List<ResourceRow> BuildRows(ResxFamily family, IReadOnlyList<ResxFile> files)
    {
        var keys = new Dictionary<string, (string Comment, Dictionary<string, string> Values)>(StringComparer.Ordinal);
        foreach (var file in files)
        {
            foreach (var entry in file.Entries)
            {
                if (!keys.TryGetValue(entry.Key, out var row))
                {
                    row = ("", new Dictionary<string, string>());
                    keys[entry.Key] = row;
                }
                row.Values[file.Locale] = entry.Value;
                if (file.Locale == ResxConstants.NeutralLocale && !string.IsNullOrEmpty(entry.Comment))
                    keys[entry.Key] = (entry.Comment, row.Values);
                else if (string.IsNullOrEmpty(row.Comment) && !string.IsNullOrEmpty(entry.Comment))
                    keys[entry.Key] = (entry.Comment, row.Values);
            }
        }

        var rows = keys.Select(kv => new ResourceRow
        {
            FamilyId = family.Id,
            Key = kv.Key,
            Comment = kv.Value.Comment,
            Values = kv.Value.Values,
            Issues = new List<ValidationIssue>()
        }).ToList();
        rows.Sort((a, b) => string.Compare(a.Key, b.Key, StringComparison.OrdinalIgnoreCase));
        return rows;
    }

    public static ExtensionSettings DefaultSettings() => new()
    {
        NeutralLocale = "",
        KeyNaming = "pascalFromNeutral",
        UpdateDesignerCs = true,
        VisibleLocales = new List<string>(),
        Rules = new ValidationRulesConfig()
    };

    private static string LocaleLabel(string locale) =>
        string.IsNullOrEmpty(locale) ? "Neutral" : locale;
}
